# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/SharedSurfacesChild.cpp
# Commit: 923dee1aab01
# Full Hash: 923dee1aab0128ea4ff60b876a8626af322084d8
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2023-12-15 05:00:55
# Description:
#   Bug 1869933 - Avoid allocating external image ID until ready for sharing with compositor process. r=gfx-reviewers,lsalzman
#   
#   There are some error paths where we could allocate an external image ID
#   for a SourceSurfaceSharedData before we are able to finish sharing it.
#   This patch makes it so that we defer allocating the ID until the last
# ==============================================================================

diff -r ef864a709ce1 -r 923dee1aab01 gfx/layers/ipc/SharedSurfacesChild.cpp
--- a/gfx/layers/ipc/SharedSurfacesChild.cpp	Thu Dec 14 23:17:37 2023 +0200
+++ b/gfx/layers/ipc/SharedSurfacesChild.cpp	Thu Dec 14 20:08:59 2023 +0000
@@ -5,7 +5,6 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "SharedSurfacesChild.h"
-#include "SharedSurfacesParent.h"
 #include "CompositorManagerChild.h"
 #include "mozilla/layers/IpcResourceUpdateQueue.h"
 #include "mozilla/layers/SourceSurfaceSharedData.h"
@@ -13,6 +12,7 @@
 #include "mozilla/layers/RenderRootStateManager.h"
 #include "mozilla/layers/WebRenderLayerManager.h"
 #include "mozilla/layers/CompositorBridgeChild.h"
+#include "mozilla/layers/CompositorManagerParent.h"
 #include "mozilla/SchedulerGroup.h"
 #include "mozilla/StaticPrefs_image.h"
 #include "mozilla/PresShell.h"
@@ -58,10 +58,9 @@
   }
 }
 
-SharedSurfacesChild::SharedUserData::SharedUserData(
-    const wr::ExternalImageId& aId)
+SharedSurfacesChild::SharedUserData::SharedUserData()
     : Runnable("SharedSurfacesChild::SharedUserData"),
-      mId(aId),
+      mId({}),
       mShared(false) {}
 
 SharedSurfacesChild::SharedUserData::~SharedUserData() {
@@ -187,17 +186,18 @@
   SharedUserData* data =
       static_cast<SharedUserData*>(aSurface->GetUserData(&sSharedKey));
   if (!data) {
-    data =
-        MakeAndAddRef<SharedUserData>(manager->GetNextExternalImageId()).take();
+    data = MakeAndAddRef<SharedUserData>().take();
     aSurface->AddUserData(&sSharedKey, data, SharedUserData::Destroy);
-  } else if (!manager->OwnsExternalImageId(data->Id())) {
+  } else if (data->IsShared()) {
+    if (manager->OwnsExternalImageId(data->Id())) {
+      // It has already been shared with the GPU process.
+      *aUserData = data;
+      return NS_OK;
+    }
+
     // If the id isn't owned by us, that means the bridge was reinitialized, due
     // to the GPU process crashing. All previous mappings have been released.
-    data->SetId(manager->GetNextExternalImageId());
-  } else if (data->IsShared()) {
-    // It has already been shared with the GPU process.
-    *aUserData = data;
-    return NS_OK;
+    data->ClearShared();
   }
 
   // Ensure that the handle doesn't get released until after we have finished
@@ -211,8 +211,8 @@
   // asking the parent instance to store a pointer to the same data, no need
   // to map the data into our memory space twice.
   if (manager->SameProcess()) {
-    SharedSurfacesParent::AddSameProcess(data->Id(), aSurface);
-    data->MarkShared();
+    data->MarkShared(manager->GetNextExternalImageId());
+    CompositorManagerParent::AddSharedSurface(data->Id(), aSurface);
     *aUserData = data;
     return NS_OK;
   }
@@ -244,7 +244,7 @@
       format == SurfaceFormat::B8G8R8X8 || format == SurfaceFormat::B8G8R8A8,
       "bad format");
 
-  data->MarkShared();
+  data->MarkShared(manager->GetNextExternalImageId());
   manager->SendAddSharedSurface(
       data->Id(),
       SurfaceDescriptorShared(aSurface->GetSize(), aSurface->Stride(), format,